aggregator/src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java

```diff
@@ -0,0 +1,205 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.cps.replay;
+
+import org.jenkinsci.plugins.workflow.cps.replay.ReplayAction;
+import org.jenkinsci.plugins.workflow.cps.replay.ReplayCause;
+import hudson.FilePath;
+import hudson.XmlFile;
+import hudson.cli.CLICommandInvoker;
+import hudson.init.InitMilestone;
+import hudson.init.Initializer;
+import hudson.model.Item;
+import hudson.model.ParametersAction;
+import hudson.model.ParametersDefinitionProperty;
+import hudson.model.Run;
+import hudson.model.StringParameterDefinition;
+import hudson.model.StringParameterValue;
+import hudson.model.User;
+import hudson.security.ACL;
+import hudson.security.GlobalMatrixAuthorizationStrategy;
+import hudson.security.Permission;
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+import jenkins.model.Jenkins;
+import jenkins.security.NotReallyRoleSensitiveCallable;
+import org.apache.commons.io.IOUtils;
+import org.hamcrest.Matchers;
+import static org.hamcrest.Matchers.containsString;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.junit.Test;
+import static org.junit.Assert.*;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class ReplayActionTest {
+
+    @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule public JenkinsRule r = new JenkinsRule();
+
+    @Test public void editSimpleDefinition() throws Exception {
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("echo 'first script'", false));
+        WorkflowRun b1 = r.assertBuildStatusSuccess(p.scheduleBuild2(0));
+        r.assertLogContains("first script", b1);
+        WorkflowRun b2;
+        /* TODO broken by ACE integration, figure out how to resurrect:
```

Comment:
Specifically with this patch

```diff
diff --git a/aggregator/src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java b/aggregator/src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java
index ffc8382..084ad86 100644
--- a/aggregator/src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java
+++ b/aggregator/src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java
@@ -24,6 +24,9 @@
 
 package org.jenkinsci.plugins.workflow.cps.replay;
 
+import com.gargoylesoftware.htmlunit.html.HtmlForm;
+import com.gargoylesoftware.htmlunit.html.HtmlPage;
+import com.gargoylesoftware.htmlunit.html.HtmlTextArea;
 import org.jenkinsci.plugins.workflow.cps.replay.ReplayAction;
 import org.jenkinsci.plugins.workflow.cps.replay.ReplayCause;
 import hudson.FilePath;
@@ -70,7 +73,6 @@ public class ReplayActionTest {
         WorkflowRun b1 = r.assertBuildStatusSuccess(p.scheduleBuild2(0));
         r.assertLogContains("first script", b1);
         WorkflowRun b2;
-        /* TODO broken by ACE integration, figure out how to resurrect:
         { // First time around, verify that UI elements are present and functional.
             ReplayAction a = b1.getAction(ReplayAction.class);
             assertNotNull(a);
@@ -87,8 +89,6 @@ public class ReplayActionTest {
             b2 = p.getBuildByNumber(2);
             assertNotNull(b2);
         } // Subsequently can use the faster whitebox method.
-        */
-        b2 = (WorkflowRun) b1.getAction(ReplayAction.class).run("echo 'second script'").get();
         r.assertLogContains("second script", r.assertBuildStatusSuccess(b2));
         ReplayCause cause = b2.getCause(ReplayCause.class);
         assertNotNull(cause);
```

the test just begins using 100% CPU:

```
"Executing editSimpleDefinition(org.jenkinsci.plugins.workflow.cps.replay.ReplayActionTest)" … waiting for monitor entry […]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$HtmlUnitContextAction.run(JavaScriptEngine.java:847)
	- waiting to lock <…> (a com.gargoylesoftware.htmlunit.html.HtmlPage)
	at net.sourceforge.htmlunit.corejs.javascript.Context.call(Context.java:628)
	at …
	at com.gargoylesoftware.htmlunit.html.HtmlInput.executeOnChangeHandlerIfAppropriate(HtmlInput.java:511)
	at com.gargoylesoftware.htmlunit.html.HtmlTextArea.setText(HtmlTextArea.java:125)
	at org.jenkinsci.plugins.workflow.cps.replay.ReplayActionTest.editSimpleDefinition(ReplayActionTest.java:84)
"JS executor for org.jvnet.hudson.test.JenkinsRule$WebClient@…" … runnable […]
   java.lang.Thread.State: RUNNABLE
	at java.util.AbstractCollection.toArray(AbstractCollection.java:139)
	at java.util.ArrayList.<init>(ArrayList.java:177)
	at com.gargoylesoftware.htmlunit.html.HtmlPage.safeGetAttributeListeners(HtmlPage.java:2097)
	- locked <…> (a java.lang.String)
	at com.gargoylesoftware.htmlunit.html.HtmlPage.fireHtmlAttributeAdded(HtmlPage.java:2060)
	at com.gargoylesoftware.htmlunit.html.HtmlElement.setAttributeNS(HtmlElement.java:203)
	at com.gargoylesoftware.htmlunit.html.DomElement.setAttribute(DomElement.java:327)
	at com.gargoylesoftware.htmlunit.javascript.host.css.CSSStyleDeclaration.writeToElement(CSSStyleDeclaration.java:696)
	at com.gargoylesoftware.htmlunit.javascript.host.css.CSSStyleDeclaration.replaceStyleAttribute(CSSStyleDeclaration.java:614)
	at com.gargoylesoftware.htmlunit.javascript.host.css.CSSStyleDeclaration.setStyleAttribute(CSSStyleDeclaration.java:586)
	at com.gargoylesoftware.htmlunit.javascript.host.css.CSSStyleDeclaration.setStyleAttribute(CSSStyleDeclaration.java:565)
	at com.gargoylesoftware.htmlunit.javascript.host.css.CSSStyleDeclaration.setStyleAttributePixel(CSSStyleDeclaration.java:6080)
	at com.gargoylesoftware.htmlunit.javascript.host.css.CSSStyleDeclaration.setHeight(CSSStyleDeclaration.java:2043)
	at sun.reflect.GeneratedMethodAccessor70.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at net.sourceforge.htmlunit.corejs.javascript.MemberBox.invoke(MemberBox.java:153)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptableObject$GetterSlot.setValue(ScriptableObject.java:311)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptableObject$RelinkedSlot.setValue(ScriptableObject.java:384)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptableObject.putImpl(ScriptableObject.java:2887)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptableObject.put(ScriptableObject.java:547)
	at com.gargoylesoftware.htmlunit.javascript.host.css.CSSStyleDeclaration.put(CSSStyleDeclaration.java:2558)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptableObject.putProperty(ScriptableObject.java:2570)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptRuntime.setObjectProp(ScriptRuntime.java:1674)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptRuntime.setObjectProp(ScriptRuntime.java:1669)
	at net.sourceforge.htmlunit.corejs.javascript.Interpreter.interpretLoop(Interpreter.java:1255)
	at net.sourceforge.htmlunit.corejs.javascript.Interpreter.interpret(Interpreter.java:798)
	at net.sourceforge.htmlunit.corejs.javascript.InterpretedFunction.call(InterpretedFunction.java:105)
	at net.sourceforge.htmlunit.corejs.javascript.ContextFactory.doTopCall(ContextFactory.java:411)
	at com.gargoylesoftware.htmlunit.javascript.HtmlUnitContextFactory.doTopCall(HtmlUnitContextFactory.java:310)
	at net.sourceforge.htmlunit.corejs.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3286)
	at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$4.doRun(JavaScriptEngine.java:789)
	at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine$HtmlUnitContextAction.run(JavaScriptEngine.java:850)
	- locked <…> (a com.gargoylesoftware.htmlunit.html.HtmlPage)
	at net.sourceforge.htmlunit.corejs.javascript.Context.call(Context.java:628)
	at net.sourceforge.htmlunit.corejs.javascript.ContextFactory.call(ContextFactory.java:513)
	at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.callFunction(JavaScriptEngine.java:796)
	at com.gargoylesoftware.htmlunit.javascript.JavaScriptEngine.callFunction(JavaScriptEngine.java:768)
	at com.gargoylesoftware.htmlunit.InteractivePage.executeJavaScriptFunctionIfPossible(InteractivePage.java:201)
	at com.gargoylesoftware.htmlunit.javascript.background.JavaScriptFunctionJob.runJavaScript(JavaScriptFunctionJob.java:53)
	at com.gargoylesoftware.htmlunit.javascript.background.JavaScriptExecutionJob.run(JavaScriptExecutionJob.java:103)
	at com.gargoylesoftware.htmlunit.javascript.background.JavaScriptJobManagerImpl.runSingleJob(JavaScriptJobManagerImpl.java:428)
	at com.gargoylesoftware.htmlunit.javascript.background.DefaultJavaScriptExecutor.run(DefaultJavaScriptExecutor.java:168)
	at java.lang.Thread.run(Thread.java:745)
```

and finally times out.

Perhaps we need a way for the `workflow-editor` form control to switch to a plain textarea when run inside HtmlUnit.